### ISSUE TITLE
Short-term action plan

### ISSUE THREAD

**Issue by cschwan** [AleCandido/partons#12]
In the past few days I thought about partons again, and I've got a few ideas to hopefully kickstart the development of it. The way I see it, we only need two things if we do it properly:

1. A Rust library and
2. a Python interface to it, using PyO3.

Here a few more concrete ideas:

- I don't think we need a management binary (LHAPDF's `lhapdf`), if we automatically download PDF sets when they are required. A good directory to cache (see #1 and #8) the sets is `${XDG_CACHE_HOME}/partons/`, and if unset `${HOME}/.cache/partons/` should be used, as recommended [here](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html). Whenever the user needs a PDF set it should be downloaded to the cache directory if it's not already there.
- In the Rust library we should support a wrapper to LHAPDF, enabled by a feature so we can have either minimal dependencies (feature off) or maximal compatibility with LHAPDF (feature on). Ideally we can select this 'backend' whenever we select a PDF set, for instance `LHAPDF:NNPDF40_nnlo_as_01180` would select NNPDF4.0 using the LHAPDF backend and `NNPDF40_nnlo_as_01180` would select the `partons`' backend. The advantage is that
   - we'll have something that will work pretty fast and
   - we can trivially test the backends against each other - it makes benchmarking very easy.
- Finally I'd like to propose a minimal interface, which is needed for `pineappl_cli`:

```
#[derive(Debug, Error)]
#[error(transparent)]
pub struct Error {}

pub type Result<T> = result::Result<T, Error>;

struct PdfUncertainty {
    pub central: f64,
    pub err_pos: f64,
    pub err_min: f64,
    pub err_sym: f64,
}

pub struct Pdf {}

impl Pdf {
    pub fn available() -> Vec<String> {}

    pub fn new(id: &str) -> Result<Self> {}

    #[must_use]
    pub fn xfx_q2(&self, id: i32, x: f64, q2: f64) -> f64 {}

    #[must_use]
    pub fn alphas_q2(&self, q2: f64) -> f64 {}

    #[must_use]
    pub fn set(&self) -> PdfSet {}

    #[must_use]
    pub fn x_min(&mut self) -> f64 {}

    #[must_use]
    pub fn x_max(&mut self) -> f64 {}

    pub fn set_force_positive(&mut self, mode: i32) {}

    #[must_use]
    pub fn force_positive(&mut self) -> i32 {}
}

pub struct PdfSet {}

pub const CL_1_SIGMA: f64 = 68.268_949_213_708_58;

impl PdfSet {
    pub fn new(id: &str) -> Result<Self> {}

    #[must_use]
    pub fn entry(&self, key: &str) -> Option<String> {}

    #[must_use]
    pub fn error_type(&self) -> String {}

    #[must_use]
    pub fn mk_pdfs(&self) -> Vec<Pdf> {}

    #[must_use]
    pub fn uncertainty(&self, values: &[f64], cl: f64) -> Result<PdfUncertainty> {}
}
```

That's basically the [`lhapdf` crate](https://crates.io/crates/lhapdf), with the methods made a bit more 'rusty'.

So my plan to get this working as quickly as possible:

1. Implement the interface above using `LHAPDF`, which is almost trivial. We'll need a bit of functionality to catch exceptions thrown when PDF sets aren't present, and in that case download them to the directory proposed above
2. Rewrite `pineappl_cli` to use this library and drop the `lhapdf` crate
3. Implement `partons`' backend - that's the difficult part.

What do you think @AleCandido? The idea here is to build the thing *top down* as quickly as possible from what we need using what we already have (LHAPDF), and later on implementing what we're initially set out to do (not relying on any shared library whose installation is complicated) rather building everything from the *bottom up*.

**Comment by alecandido** [comment]
> 1. A Rust library and
> 2. a Python interface to it, using PyO3.

This part of the template is already in place. The moment we have content for Rust, I will quickly bind to Python mirroring PineAPPL (for simplicity, we re-apply a proven pattern, improvements later on).

> - I don't think we need a management binary (LHAPDF's `lhapdf`), if we automatically download PDF sets when they are required. A good directory to cache (see #1 and #8) the sets is `${XDG_CACHE_HOME}/partons/`, and if unset `${HOME}/.cache/partons/` should be used, as recommended [here](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html). Whenever the user needs a PDF set it should be downloaded to the cache directory if it's not already there.

Indeed, this is where I started from:
0. I defined an interface for a PDF repo; I made sure to allow enough flexibility to support https://lhapdf.hepforge.org/pdfsets.html, but I tried not to have much more, to keep it simple (in principle, I want to support multiple sources, such that we can install NNPDF sets from our own server, even if it is not official on LHAPDF)
1. I started fetching info files (done)
2. and found out they are terribly non-standard (so I'm stuck at parsing them, because there are a few ones so bad that I can't parse with `serde-yaml`...)
3. next step would have been to fetch the tarball (maybe done)
4. and parse its content (I already have a simple parser in Python, it takes less than one hour to translate)

The actually missing part is that I decided I do not want to depend on GSL, so I wanted to reimplement LHAPDF native interpolation in Rust. It has the farther advantage of being completely non-breaking with LHAPDF.
 It is a translation, and files are not so long (~4 files, ~100 lines each):
- https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/KnotArray.cc
- https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/Interpolator.cc
- https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/LogBicubicInterpolator.cc
- https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/LogBilinearInterpolator.cc

and I would put into `ndinterp`.
In this way, `partons` will be fully standalone, with no system deps.

Same thing will have to be done for $\alpha_S$, but it is even simpler (1-d), and in any case one thing at a time.

About the cache directories and related, I'm fully aware of XDG, and great fan of it. But here I found a reliable library, that is taking care of it for me. It applies XDG standard for Linux, but it does also take care of MacOS and Windows standards:
https://github.com/AleCandido/partons/blob/9a4ff1543c9a358b8e5233bcc81b18dc0934e44d/partons/Cargo.toml#L15
https://github.com/AleCandido/partons/blob/9a4ff1543c9a358b8e5233bcc81b18dc0934e44d/partons/src/configs.rs#L53-L59

**Comment by alecandido** [comment]
> - In the Rust library we should support a wrapper to LHAPDF, enabled by a feature so we can have either minimal dependencies (feature off) or maximal compatibility with LHAPDF (feature on). Ideally we can select this 'backend' whenever we select a PDF set, for instance `LHAPDF:NNPDF40_nnlo_as_01180` would select NNPDF4.0 using the LHAPDF backend and `NNPDF40_nnlo_as_01180` would select the `partons`' backend. The advantage is that
>   - we'll have something that will work pretty fast and
>   - we can trivially test the backends against each other - it makes benchmarking very easy.

This is an interesting idea, and I'd write a separate issue.
Personally, I would not focus on wrapping LHAPDF immediately, because I find it more cumbersome (still requires an LHAPDF installation) and I'm trying to preserve compatibility in the `partons` backend as well. But if you want to work on it, feel free to start as soon as you like.

**Comment by alecandido** [comment]
> - Finally I'd like to propose a minimal interface, which is needed for `pineappl_cli`:

This is perfectly fine.
And we can even release that interface at the beginning, and upgrade it later on (if ever needed). PineAPPL can upgrade to the new release the moment is available, or stick to the initial version for as long as it is needed.

**Comment by alecandido** [comment]
> So my plan to get this working as quickly as possible:

> 1. Implement the interface above using `LHAPDF`, which is almost trivial. We'll need a bit of functionality to catch exceptions thrown when PDF sets aren't present, and in that case download them to the directory proposed above
> 2. Rewrite `pineappl_cli` to use this library and drop the `lhapdf` crate
> 3. Implement `partons`' backend - that's the difficult part.

I was focusing on part 3., also because I felt simpler for me. However, if you want to start with part 1. I can also help with that, but I'd suggest that you lead the effort, since you have clearer than me what to do (you already did once :wink:).
Once that point 1. is done, I believe 2. should be rather quick, since it will be just a Rustier version of `lhapdf` crate.

**Comment by alecandido** [comment]
Just to mention it explicitly: making a service to PineAPPL was one of the main goals. But, in the end, it is one and the same thing of having a proper package that is easier to distribute, and thus easier to install for users. Both Rust and Python users.

**Comment by scarlehoff** [comment]
> The actually missing part is that I decided I do not want to depend on GSL, so I wanted to reimplement LHAPDF native interpolation in Rust. It has the farther advantage of being completely non-breaking with LHAPDF.
> It is a translation, and files are not so long (~4 files, ~100 lines each):
> 
> https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/KnotArray.cc
> https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/Interpolator.cc
> https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/LogBicubicInterpolator.cc
> https://gitlab.com/hepcedar/lhapdf/-/blob/main/src/LogBilinearInterpolator.cc
> and I would put into ndinterp.
> In this way, partons will be fully standalone, with no system deps.

I call dibs on this.

**Comment by cschwan** [comment]
> About the cache directories and related, I'm fully aware of XDG, and great fan of it. But here I found a reliable library, that is taking care of it for me. It applies XDG standard for Linux, but it does also take care of MacOS and Windows standards:
> 
> https://github.com/AleCandido/partons/blob/9a4ff1543c9a358b8e5233bcc81b18dc0934e44d/partons/Cargo.toml#L15

I wasn't aware of this, that's very useful!

**Comment by alecandido** [comment]
> I wasn't aware of this, that's very useful!

It also has a "lower-level" companion: https://github.com/dirs-dev/dirs-rs

And, only now I realize even a CLI https://github.com/dirs-dev/dirs-cli-rs (+ other projects less interesting for me: .NET, JVM, sys...)

**Comment by cschwan** [comment]
I'm closing this Issue, the remaining points are addressed in #13.
